bdio-protobuf/src/main/java/com/blackducksoftware/bdio/proto/ProtobufBdioReader.java

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright (C) 2023 Synopsys Inc.
+ * http://www.synopsys.com/
+ * All rights reserved.
+ *
+ * This software is the confidential and proprietary information of
+ * Synopsys ("Confidential Information"). You shall not
+ * disclose such Confidential Information and shall use it only in
+ * accordance with the terms of the license agreement you entered into
+ * with Synopsys.
+ */
+package com.blackducksoftware.bdio.proto;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.zip.ZipInputStream;
+
+import com.blackducksoftware.bdio.proto.v1.ProtoAnnotationNode;
+import com.blackducksoftware.bdio.proto.v1.ProtoComponentNode;
+import com.blackducksoftware.bdio.proto.v1.ProtoDependencyNode;
+import com.blackducksoftware.bdio.proto.v1.ProtoFileNode;
+import com.blackducksoftware.bdio.proto.v1.ProtoImageLayerNode;
+import com.blackducksoftware.bdio.proto.v1.ProtoImageNode;
+import com.blackducksoftware.bdio.proto.v1.ProtoScanHeader;
+import com.google.common.primitives.Shorts;
+import com.google.protobuf.Any;
+
+/**
+ * Utility class for reading and deserializing bdio data in protobuf format for
+ * different scan types
+ *
+ * @author sharapov
+ * @author johara
+ */
+public class ProtobufBdioReader {
+
+	/**
+	 * Returns true if given input stream represents header chunk
+	 *
+	 * @param in input stream
+	 * @return true if given input stream represents header chunk
+	 * @throws IOException
+	 */
+	public static boolean isHeader(InputStream in) throws IOException {
+		byte[] messageTypeBytes = new byte[2];
+		in.read(messageTypeBytes);
```

Comment:
I don't know what the underlying implementation is but this will read to the length of the provided byte array for certain implementations like FileInputStream.   I'm not sure this holds or is as safe to assume for other input stream implementations.   There are some similar patterns of reading below.

Reply:
Currently, there are three implementations of inputstream, that we use to read data.
1. BufferedInputStream backed by FileInputStream (read chunk in scan container)
2. ByteArrayInputStream (read separate chunks coming from database)
3. ZipInputStream (read data from bdio archive, when single bdio archive is uploaded or when it is fetched from db)

All of them work fine. I agree that there is no guarantee that arbitrary implementation of InputStream.read(byte[]) method  follows description of how this method should work (provided by the javadoc of this method).
To make things more safe we can narrow the type, that is instead of InputStream as the argument, we could have BufferedInputStream. At the same time, if there will be a problem with some particular implementation of InputStream that will be used in the future, it will be pretty easy to find the issue during the testing, and we can handle the problem at that time.
What do you think?